src/pipeline.cpp

```diff
@@ -387,27 +397,25 @@ static void buildEarlyOptimizerPipeline(ModulePassManager &MPM, PassBuilder *PB,
       if (O.getSpeedupLevel() >= 1) {
           FunctionPassManager FPM;
           if (O.getSpeedupLevel() >= 2) {
-#if JL_LLVM_VERSION >= 160000
-              // TODO check the LLVM 15 default.
-              FPM.addPass(SROAPass(SROAOptions::PreserveCFG));
-#else
-              FPM.addPass(SROAPass());
-#endif
-              // SROA can duplicate PHI nodes which can block LowerSIMD
-              FPM.addPass(InstCombinePass());
-              FPM.addPass(JumpThreadingPass());
-              FPM.addPass(CorrelatedValuePropagationPass());
-              FPM.addPass(ReassociatePass());
-              FPM.addPass(EarlyCSEPass());
-              JULIA_PASS(FPM.addPass(AllocOptPass()));
-          } else { // if (O.getSpeedupLevel() >= 1) (exactly)
-              FPM.addPass(InstCombinePass());
-              FPM.addPass(EarlyCSEPass());
-          }
-          invokePeepholeEPCallbacks(FPM, PB, O);
-          MPM.addPass(createModuleToFunctionPassAdaptor(std::move(FPM)));
-      }
-      MPM.addPass(GlobalDCEPass());
+            FPM.addPass(SROAPass(SROAOptions::ModifyCFG));
+            FPM.addPass(EarlyCSEPass(true));
+            FPM.addPass(JumpThreadingPass());
+            FPM.addPass(CorrelatedValuePropagationPass());
+            FPM.addPass(SimplifyCFGPass(basicSimplifyCFGOptions()));
+            FPM.addPass(InstCombinePass());
+            FPM.addPass(AggressiveInstCombinePass());
```

Comment:
does it make sense to do 2 instcombine right next to each other?

Reply:
Yep

Reply:
It might be worth customizing the `AggressiveInstCombinePass` slightly since the defaults include some options that are likely not useful for us specifically from https://llvm.org/doxygen/AggressiveInstCombine_8cpp.html:
* `foldSqrt` is probalby useless because we generate LLVM sqrt
* `tryToRecognizePopCount` probably isn't useful since we have `count_ones`
* `foldMemChr` I don't think we use `memchr` (but not sure).

This is unlikely to matter much, but probably could save a bit of compile time here and there.